TACC/Core-Portal
# Task/fp 506 dropdown selector

## Overview: ##

Create a `DropdownSelector` component.

## PR Status: ##


## Related Jira tickets: ##

* [FP-506](https://jira.tacc.utexas.edu/browse/FP-506)

## Summary of Changes: ##

- _Minor_: Prevent CSS lint error on `composes` (also available in #96)
- __New__: Add `DropdownSelector`\* component
- __New__: Add `UIPatternsDropdownSelector` to "UI Patterns" section
-  _Minor_: Rename `UIPatternsMessages.js` → `UIPatternsMessage.js`.
-  _Minor_: Add `bootstrap.form.css` (to isolate re-usable overrides to `.form-control`) (also available in #96)
-  _Minor_: Add `external.select-css.css` (to isolate and reference styles necessary for `<select>`'s)

\* The name is overly explicit, but for these reasons: (1) "dropdown" is vague (can be a menu, or a `<select>` field), but is, unfortunately, a common name for this element; (2) as agreed by design, any widget that performs an action (and does not have an existing common/standard industry name) must be an "-er" or "-or" noun (like "File Uploader" and "File Selector").

## Testing Steps: ##
1. Load "UI Patterns" in different browsers.
2. Compare the field (not the dropdown) of the "Default" (`single`) "DropdownSelector".
3. See "Known Issues".

## UI Photos:

### Field (Chrome, Firefox)
<img width="309" alt="Chrome" src="https://user-images.githubusercontent.com/62723358/88867900-5e161980-d1d4-11ea-9da1-494c66089c6f.png">
<img width="309" alt="Firefox" src="https://user-images.githubusercontent.com/62723358/88867903-5eaeb000-d1d4-11ea-8852-d63d04ddf8e3.png">

### Other Form Controls Not Broken

Actually fixes a design bug, border radius removed on fields:
[Global form-control Overrides Do Not Break App Form.zip](https://github.com/TACC/Frontera-Portal/files/5003828/Global.form-control.Overrides.Do.Not.Break.App.Form.zip)

(Except Safari's, but it's `<select>` was already wild and I will tame later in a ticket specific to normal form UI.)

### Field, Menu, Select Multiple, Focus (all browsers)
[DropdownSelector Screenshots.zip](https://github.com/TACC/Frontera-Portal/files/4998165/DropdownSelector.Screenshots.zip)

## Notes: ##

### Known Issues

Styling all aspects of a native `<select>` element in a way that matches cross-browser is **not** possible¹. _This has been and still is a UI black hole for decades._

1. **All**: The menus have unique styles per browser and/or operating system.
2. **Safari**: The element cascades certain styles (`color`, `font-…`) to menu texts, and they cannot be [`revert`](https://developer.mozilla.org/en-US/docs/Web/CSS/revert)-ed nor [`initial`](https://developer.mozilla.org/en-US/docs/Web/CSS/initial)-ed nor [`unset`](https://developer.mozilla.org/en-US/docs/Web/CSS/unset).
3. **Firefox**: A `width: auto` on the field implies a min-width value equal to the longest `optgroup` (in other browsers, the value is is equal to the longest `option`).
4. **Firefox**: Extra horizontal space in field. Options: [ignore](https://stackoverflow.com/a/27006710/11817077), [hack-a-lot](https://stackoverflow.com/a/58783503/11817077), use plugin.

The alternative is to use a plugin. I know one that is very well-written and very powerful, but it is overkill for our _current_ usage. Researching a good `<select>` plugin takes more time than is available; I knew this before I wrote this task, which is why I, ahead of time, discussed with Design—they have accepted differences caused by issue #1\*. Issues #2 and #3 are not a priority, because they are irrelevant for the tasks that use this component ([FP-508](https://jira.tacc.utexas.edu/browse/FP-508), [FP-218](https://jira.tacc.utexas.edu/browse/FP-218)). Issue #4 exists because I do not want to implement browser hacks.

\* See "References" at [Shared UI - Components - (Select) Dropdown Field](https://confluence.tacc.utexas.edu/x/BoDQCQ).
